### DISCUSSION TITLE
expanding a view based on a referenced collection

### DISCUSSION THREAD

**Discussion by RobinNaumann** [pocketbase/pocketbase#4188] [Q&A] [answered]
Hey there,

I'm currently developing a small app using pocketbase. Right now I'm facing the following issue (which I could not find an answer to online)

suppose I have the following two collections:
```yaml
account
    id
    name
    email
```

```yaml
post
    id
    author   (relation field to account)
    ...
```

Now, I want other users to see the name of the `author` of a `post`, **but not her `email`**. To achieve this, I created a view collection:

```yaml
profile
   SELECT id, name FROM account
```

To reduce network trafic, I now want to retrieve all `post` related information in one request. This is what I came up with:

```ts
(id:string) => collectionPost.getOne(id, {expand: "author.profile(id_link)"});
```

This works, but: 
 - of course, requires access to the `Account` collection. (as it also expands `{author: account{...}}`
 - requires another (non key) id field in `profile` (e.g. `SELECT id, id as id_link, name FROM account`)
 
My question is now: **How can I expand `profile` into my `post` without needing view access to `account`.**

Thank you in advance,
Robin

*PS: As I understand it, I could create a `post_view` view collection and reference `profile` from there, but this introduces potentials for inconsistencies (for example when editing the fields in `post` while forgetting to edit the fields in `post_view`)*

**Comment by ganigeorgiev** [comment, marked as answer]
Note that the user email will be visible by:
- admins
- by the owner of the user record
- by everyone **only** if the user decided to make their email public (aka. `emailVisibility` flag is `true`)

> How can I expand profile into my post without needing view access to account

If you still want to use a separate view collection, the only way I can think of at the moment is to change your view query to include a relation field back to the `post` collection. Something like:

```sql
SELECT
    (ROW_NUMBER() OVER()) as id
    account.name,
    p.id as post
FROM account
INNER JOIN post p on p.author = account.id
```
Then your `expand` could look something like:
```js
expand: "profile(post)"
```

_Side-note: for optimization you can define an index on the relation `author` field from the "Unique constraint and indexes" section below the collection fields (in future this will be either automated or have a note/suggestion in the Admin UI)._

**Reply by RobinNaumann** [reply]
Hi,

thanks for the quick answer! Yeah, I thought about that, but this would require adding a _"back-relation"_ to each collection I want to reference `profile` from. This is fine, I was just wondering if there was maybe a more generalizable approach.

in my mind, something in the direction of this would be great:

`expand: profile(id = author)` 

*(where `id` (or perhaps a non-key `id_link`) is `profile.id` and `author` is `post.author`)*

But yeah, I have no idea which side effects this might have. 

I really enjoy using pocketbase, my hat goes off to you!
all the best, 
Robin

**Reply by ganigeorgiev** [reply]
No, this is not possible at the moment. An alternative could be to move any private nonsystem `account` fields to a separate collection with its own API rules (eg. `profile`) and `expand` safely directly the `account` collection (aka. without a view). Again, note that the email will be publicly visible by other users only if the user decide to make their email public.

**Reply by RobinNaumann** [reply]
I understand. Yeah besides hiding the email, the profile view also calculates and returns a few additional derived values. I'll just try the "back-relation" approach you outlined above. thanks; Robin
